### PR TITLE
Enable to change animation speed

### LINKED ISSUE #3125: Change animation speed
### Discussed in https://github.com/google/model-viewer/discussions/2795

<div type='discussions-op-text'>

<sup>Originally posted by **cw789** September 16, 2021</sup>
Hello everybody.

Is there a possibility to change the animation speed during the play?</div>

@elalish suggests using requestAnimationFrame, but there is `timeScale` in Three.js ( `AnimationAction` or `AnimationMixer`) that may allow you to change the animation speed.
Are there any plans to add `currentTimeScale` as property ?

### PR DESCRIPTION
**Pull request by futahei**
### Reference Issue
Fixes #3125 

### What I do
- add `timeScale` property
  - You can change animation speed. If you set to a negative value (ex. `-2`), the animation play reverse.
- fix last frame position when the animation played a specified number of time
  - Always stop the animation at the last frame when it ends.

### CODE REVIEW

File: packages/model-viewer/src/features/animation.ts [hunk 1]
```diff
@@ -101,6 +101,16 @@ export const AnimationMixin = <T extends Constructor<ModelViewerElementBase>>(
...
     }
 
+    get currentTimeScale(): number {
+      return this[$scene].animationTimeScale;
+    }
+
+    set currentTimeScale(value: number) {
+      this[$scene].animationTimeScale = value;
+      this[$renderer].threeRenderer.shadowMap.needsUpdate = true;
+      this[$needsRender]();
```
Comment (elalish) [inline packages/model-viewer/src/features/animation.ts:111]: Are these renderer updates necessary? If the animation is playing, these are being called anyway, and if not, then changing the time scale shouldn't affect the  render, right?

File: packages/model-viewer/src/three-components/ModelScene.ts [hunk 2]
```diff
@@ -556,6 +564,7 @@ export class ModelScene extends Scene {
       action.setLoop(loopMode, repetitionCount);
 
       action.enabled = true;
+      action.clampWhenFinished = true;
```
Comment (elalish) [inline packages/model-viewer/src/three-components/ModelScene.ts:567]: 👍

File: packages/model-viewer/src/features/animation.ts [hunk 3]
```diff
@@ -41,6 +41,7 @@ export declare interface AnimationInterface {
   readonly paused: boolean;
   readonly duration: number;
   currentTime: number;
+  currentTimeScale: number;
```
Comment (elalish) [inline packages/model-viewer/src/features/animation.ts:44]: maybe just `timeScale`?

File: packages/model-viewer/src/features/animation.ts [hunk 4]
```diff
@@ -101,6 +101,16 @@ export const AnimationMixin = <T extends Constructor<ModelViewerElementBase>>(
       this[$needsRender]();
     }
 
+    get currentTimeScale(): number {
```
Comment (elalish) [inline packages/model-viewer/src/features/animation.ts:104]: Could you add this to the docs and perhaps add/modify an animation example to demonstrate this?

**Review by elalish — CHANGES_REQUESTED** [review]
Thanks! Just a few things...

File: packages/model-viewer/src/features/animation.ts (hunk 3 above)
Comment (futahei) [reply packages/model-viewer/src/features/animation.ts:44]: I was aware of the `currentTime`, but you're right.

File: packages/model-viewer/src/features/animation.ts (hunk 1 above)
Comment (futahei) [reply packages/model-viewer/src/features/animation.ts:111]: fix it

File: packages/model-viewer/src/features/animation.ts (hunk 4 above)
Comment (futahei) [reply packages/model-viewer/src/features/animation.ts]: added docs & sample

**Review by elalish — APPROVED** [review]
Thanks, and nice example!

**Comment by futahei** [comment]
@elalish Thank you for merging it.
I use model-viewer in my products, and the fix I made is very useful. Do you have any plans to release the next version?

**Comment by elalish** [comment]
@futahei Thanks for contributing! Ideally I try to  release every couple of months, though it varies a lot. If you need a feature sooner, I'd recommend you build locally and serve it yourself instead of using unpkg. Of course I'd also recommend you do some extra testing, but we generally try to keep master pretty stable.

**Comment by Jayemby** [comment]
I wanted to raise this as an issue but the template terrified me.

I've just (23/02/22) done a local build to get the `timeScale` property to work (brilliant work, by the way). However, the `repetitions` option set to 1 in `play()` is broken for negative `timeScale`s

to recreate:
optionally, add a listener / bind a debug function to the `finished` event.
then:
```
modelViewer.timescale = -1;
modelViewer.play({repetitions:1});
```
observe that no animation plays (and that `finished` is triggered immediately triggered). I even tried -1 repetitions but that didn't work either.
to be clear: `modelViewer.play({repetitions:x});` still works like a charm for every positive `x` other than 1 for a negative `timeScale`. In fact, a workaround for this is to use `modelViewer.play({repetitions:0.9});` as it actually gives the desired affect of `modelViewer.play({repetitions:1});`.

**Comment by elalish** [comment]
@Jayemby can you tell me what about the template terrified you? A new issue is definitely a better way to report a bug. And if you remix our Glitch to demonstrate your issue we'll be able to find and fix it much faster.

**Comment by Jayemby** [comment]
Sorry, was mainly because the Glitch wouldn't load so I thought I was doing something wrong; I've not used Glitch before. The tone of the template was very imperative and said I should open a Q&A if I didn't have the info to fill out the template completely. I didn't want to get yelled at by a stranger on the internet 😅

I've managed to put something together and I'm going to open an issue

**Comment by elalish** [comment]
Thanks for the feedback; I may need to tone it down a shade. We just had a *lot* of people deleting the template and asking a random question.

**Comment by Jayemby** [comment]
maybe not; I might just be too insecure lol

**Pull request merged**